http_server/include/HttpServer.hpp

```diff
@@ -0,0 +1,387 @@
+
+            auto body = receiveBody<SIZE_BODY>(req);
+            if (!body) {
+                if (body.error() == ESP_ERR_NO_MEM) {
+                    httpd_resp_send_custom_err(req, statusCodePhrase(StatusCode::PayloadTooLarge), statusCodePhrase(StatusCode::PayloadTooLarge));
+                } else if (body.error() == HTTPD_SOCK_ERR_TIMEOUT) {
+                    httpd_resp_send_408(req);
+                }
+                return body.error();
+            }
+
+            for (const auto& header: data->headers) {
+                auto err = httpd_resp_set_hdr(req, header.first.c_str(), header.second.c_str());
+                assert(err == ESP_OK && "Error setting headers");
+            }
+
+            auto parameters = getQueryParameters(req, data->parameters);
+
+            auto ret = data->callback(req->uri, body.value(), parameters, data->userContext);
+            if (!ret.has_value()) {
+                ESP_LOGD(TAG, "Callback for %s returned an error: %s", req->uri, ret.error().statusMessage());
+                httpd_resp_send_custom_err(req, ret.error().statusMessage(), ret.error().message.c_str());
+                return ESP_FAIL;
+            }
+
+            httpd_resp_send(req, ret.value().c_str(), ret.value().length());
+            return ESP_OK;
+        }
+
+        /**
+         * @brief Get the query parameters from the request uri. Also decodes the parameters
+         * @tparam N_PARAMETERS The number of parameters to get.
+         * @param req Pointer to the request object that was made.
+         * @param keys The keys to get the parameters for.
+         * @return A map of the found query parameters. If a key is not found, it will not be in the map.
+         */
+        template<std::size_t N_PARAMETERS>
+        static etl::unordered_map<QueryKey, QueryValue, N_PARAMETERS> getQueryParameters(httpd_req_t* req, const std::array<QueryKey, N_PARAMETERS>& keys) {
+            etl::string<QueryKey::MAX_SIZE + QueryValue::MAX_SIZE> query;
+            auto                                           err = httpd_req_get_url_query_str(req, query.data(), query.max_size());
+            assert((err == ESP_OK || err == ESP_ERR_NOT_FOUND) && "Error getting URL query string");
+            // If there are no query parameters, return an empty map
+            if (err == ESP_ERR_NOT_FOUND) {
+                return {};
+            }
+
+            etl::unordered_map<QueryKey, QueryValue, N_PARAMETERS> parameters;
+            for (const auto& key: keys) {
+                char buf[QueryValue::MAX_SIZE];
+                err = httpd_query_key_value(query.c_str(), key.c_str(), buf, sizeof(buf));
+                assert((err == ESP_OK || err == ESP_ERR_NOT_FOUND) && "Error getting query key value");
+                // If the key is not found, continue to the next key
+                if (err == ESP_ERR_NOT_FOUND) {
+                    continue;
```

Comment:
perhaps add a warning log here to help with future programming

Reply:
The httpd lib already logs warnings on its own when that happens